src/java.xml/share/classes/javax/xml/catalog/Util.java

```diff
@@ -267,4 +269,34 @@ static void validateFeatureInput(Feature f, String value) {
             Util.validateUrisSyntax(value.split(";"));
         }
     }
+
+    /**
+     * Returns the absolute form of the specified uri after resolving it against
+     * the base. Returns the uri as is if it's already absolute.
+     *
+     * @param base the base, that is the system id of the catalog within the
+     * Catalog implementation
+     * @param uri the specified uri
+     * @return Returns the absolute form of the specified uri
+     */
+    static String getAbsoluteURI(String base, String uri) {
+        try {
+            URI baseURI = URI.create(base);
+            URI specURI = URI.create(uri);
+            String temp;
+            if (specURI.isAbsolute()) {
+                temp = specURI.toURL().toString();
+            } else {
+                if (SCHEME_JAR.equalsIgnoreCase(baseURI.getScheme())) {
+                    temp = Paths.get(base, uri).toString();
+                } else {
+                    temp = baseURI.resolve(uri).toString();
+                }
+            }
+            return temp;
+        } catch (MalformedURLException ex) {
+            // no action, shouldn't happen as the base has already been validated
```

Comment:
If this "shouldn't happen", is that considered an Error, which should be reported?